sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -376,16 +389,11 @@ public void finalizeCheckpoint() throws IOException {
       this.source = source;
       this.current = null;
       this.checkpointMark = checkpointMark != null ? checkpointMark : new RabbitMQCheckpointMark();
-      try {
-        connectionHandler = new ConnectionHandler(source.spec.uri());
-      } catch (Exception e) {
-        throw new IOException(e);
-      }
     }
 
     @Override
     public Instant getWatermark() {
-      return checkpointMark.oldestTimestamp;
+      return checkpointMark.latestTimestamp;
```

Comment:
Ideally, the watermark would be informed by the RabbitMQ broker's backlog. The goal is for the watermark to be a good prediction of the fact "no future records will have timestamps before the watermark's timestamp".

My impression is that you have made the watermark advance more reliably, in good ways, but will perhaps result in a large amount of data being marked late. What do you think?

Reply:
I'm not sure I follow. 

>  The goal is for the watermark to be a good prediction of the fact "no future records will have timestamps before the watermark's timestamp".

In this implementation, the timestamp is effectively processing time. The timestamp is the delivery timestamp (which is how this was implemented even before this PR), which I believe is when the consumer picked up the message, not when it was enqueued. The queue is FIFO (we should probably add explicit documentation that priority queues are unsupported).

As noted in the PR description, the AmqpIO and certain Kafka timestamp watermarking implementations directly use 'now', so there's prior art here.

> when there's not a lot of messages (but in that case, we can imagine to update the watermark before the finalize checkpoint

This use case is what prompted this PR originally. 

> when we use an existing exchange (instead of creating one)

I don't see how this is relevant. Messages are pulled from a single queue, regardless of how the exchange was declared. 

> As it depends of the use case